[17.0][MIG] purchase_triple_discount: Migration to 17.0

ramiadavid opened the pull request:
Updated to work with standard purchase discount added in odoo 17.
Updated to use https://github.com/OCA/account-invoicing/pull/1638 method

As I see it, it maintains all the functions that existed until now except:

- default_supplierinfo_discount field in the supplier.
- Purchase report does not have discount fields
- Purchase PDF document does not have discount fields

This is because in the Odoo standard the discount field does not have these functions, and it did not seem logical to me to add it to the other two if the first one does not have it.

The purchase_discount https://github.com/OCA/purchase-workflow/pull/2114 module can be omitted, but if you want to add that functions, I think it would be better to create a new modules for this purpose.

Depends on:

- [ ] https://github.com/OCA/account-invoicing/pull/1626

chienandalu committed:
[ADD] purchase_triple_discount: New module

pedrobaeza committed:
[FIX] purchase_triple_discount: Correct view inheritance

oca-transbot committed:
OCA Transbot updated translations from Transifex

chienandalu committed:
[MIG] purchase_triple_discount: Migration to 11.0

legalsylvain committed:
[MIG] purchase_triple_discount

- remove dependency to obsolete module product_supplierinfo_discount
- rename procurement.rule into stock.rule
- Use new hook introduced in purchase_discount V12 to add discount2 and discount3
- Introduce discount2 and discount3 in purchase.report model

legalsylvain committed:
[IMP] inheritance between purchase_discount and purchase_triple_discount, adding hook

pedrobaeza committed:
[IMP] purchase*_discount: Several improvements:

* Lighter algorithm when synchronizing PO line info to product.supplierinfo.
* Use new create multi method for performance.
* Better SQL expresion in purchase report for decreasing the number of operations.
* Add test for purchase report.

oca-travis committed:
[UPD] Update purchase_triple_discount.pot

OCA-git-bot committed:
purchase_triple_discount 12.0.1.0.1

OCA-git-bot committed:
[UPD] README.rst

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/

liweijie0812 committed:
Added translation using Weblate (Chinese (Simplified))

liweijie0812 committed:
Translated using Weblate (Chinese (Simplified))

Currently translated at 100.0% (15 of 15 strings)

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/zh_CN/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/

HviorForgeFlow committed:
[FIX] purchase_triple_discount: Missing return raises error

OCA-git-bot committed:
purchase_triple_discount 12.0.1.0.2

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (15 of 15 strings)

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/it/

legalsylvain committed:
[FIX] purchase_triple_discount: bad display of discounts 2 and 3 in supplierinfo form view

OCA-git-bot committed:
purchase_triple_discount 12.0.1.1.0

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 20.0% (3 of 15 strings)

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/sl/

marcelsavegnago committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 20.0% (3 of 15 strings)

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/pt_BR/

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/

sysadminmatmoz committed:
Translated using Weblate (Slovenian)

Currently translated at 100.0% (15 of 15 strings)

Translation: purchase-workflow-12.0/purchase-workflow-12.0-purchase_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/purchase-workflow-12-0/purchase-workflow-12-0-purchase_triple_discount/sl/

ChrisOForgeFlow committed:
[IMP] purchase_triple_discount: black, isort, prettier

ChrisOForgeFlow committed:
[14.0][MIG] purchase_triple_discount

oca-travis committed:
[UPD] Update purchase_triple_discount.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[IMP] update dotfiles [ci skip]

MiquelRForgeFlow committed:
[FIX] purchase_triple_discount: decimal_precision is obsolete

OCA-git-bot committed:
purchase_triple_discount 14.0.1.0.1

ramiadavid committed:
[MIG] purchase_triple_discount: Migration to 15.0

ao-landoo committed:
[IMP] purchase_triple_discount: Display discount2 and discount3 fields on purchase report

(cherry picked from commit 6e5b6527c5fa34878b007b473b709143e79c3f36)

Commit:
[UPD] Update purchase_triple_discount.pot

OCA-git-bot committed:
[UPD] README.rst

ramiadavid committed:
[MIG] purchase_triple_discount: Migration to 16.0

ramiadavid committed:
[IMP] purchase_triple_discount: pre-commit auto fixes

pedrobaeza commented:
/ocabot migration purchase_triple_discount

FernandoRomera commented on purchase_triple_discount/models/product_supplierinfo.py:
```diff
@@ -0,0 +1,80 @@
+        ),
+    ]
+
+    @api.depends("discount1", "discount2", "discount3")
```
Suggestion
@api.depends(lambda self: self._get_multiple_discount_field_names())

ramiadavid replied:
I had never used it like this, I didn't know it was possible.
I have already changed it, but I have made other changes:
  - I have moved all the code that was repeated in purchase.order.line and in purchase.supplierinfo to an AbstractModel, to avoid repeating code
  - All functions that obtained the values of dicount1 , discount2, etc... use _get_multiple_discount_field_names() to obtain the fields involved

FernandoRomera commented on purchase_triple_discount/models/purchase_triple_discount_mixin.py:
```diff
@@ -0,0 +1,79 @@
+        store=True,
+    )
+    discount1 = fields.Float(
+        string="Disc. (%)",
```
Original:
discount = fields.Float(string="Discount (%)", digits="Discount")

Suggestion:
discount1 = fields.Float(string="Disc. 1 (%)", digits="Discount")

Add number one.

ramiadavid replied:
Done

ramiadavid committed:
[REF] purchase_triple_discount: Consolidate discount in std field

Until now, the triple discount feature did use the standard Odoo field
as the first discount field, adding only extra fields for the second
and the third discount. This implied we had to override any function
using discount to consider the other discounts properly.

By adding an extra field discount1 to store the first discount, it allows
to redefine the standard discount field to a computed field that will
consolidate the triple discount from the other fields, and avoid the
need to redefine anything relying on the discount field as it will
already consider the triple discounts.

pedrobaeza committed:
[IMP] purchase_discount/purchase_triple_discount: Use core hook

Following https://github.com/odoo/odoo/pull/175998, we can remove the
hacks done for cleaner code.

OCA-git-bot committed:
[BOT] post-merge updates

ValentinVinagre committed:
[FIX] purchase_discount: typo error in post-migrate script

ramiadavid committed:
[MIG] purchase_triple_discount: Migration to 17.0

ivilata commented:
This PR depends on OCA/account-invoicing#1626, which has been idle for a while (although it counts with many positive reviews and passes tests); in the meanwhile it was superseded by OCA/account-invoicing#2114, which includes some refactorings and passes tests, but still lacks reviews. I'll try to call the attention of some of the former PR reviewers on the new one.

ivilata commented:
@ramiadavid, OCA/account-invoicing#2114 has been merged, and it may fix the failing tests because of `account_invoice_triple_discount`. Would you mind rebasing and pushing so as to trigger tests again? We may be on the good track to finally get this done! :smiley: 

Thanks again for your PR!

Reyes4711-S73 reviewed (approved):
LGTM

Segui-S73 reviewed (approved):
LGTM

Reyes4711-S73 commented:
@OCA/procurement-purchase-maintainers Please, can you merge this PR?

ivilata commented:
Thanks @ramiadavid for the push! I see that both checks "test with OCB" and "test with Odoo" do contain an error which may be relevant (both on line 503 in the last run):

```
2025-11-08 08:41:09,134 316 ERROR odoo odoo.addons.purchase_triple_discount.tests.test_purchase_discount: FAIL: TestPurchaseOrder.test_07_supplierinfo_from_purchaseorder
Traceback (most recent call last):
  File "/__w/purchase-workflow/purchase-workflow/purchase_triple_discount/tests/test_purchase_discount.py", line 212, in test_07_supplierinfo_from_purchaseorder
    self.assertEqual(seller.discount1, 11.11)
AssertionError: 53.910000000000004 != 11.11
```

ramiadavid commented:
@ivilata Yes, I'm looking into this issue.

ramiadavid commented:
@ivilata  done

ivilata commented:
Thanks @ramiadavid! @pedrobaeza (since you tagged this PR a while ago), maybe this can be finally merged? :slightly_smiling_face: Thank you!

pedrobaeza commented:
/ocabot merge nobump

ivilata commented:
Thanks a lot @ramiadavid (and @pedrobaeza)! :pray:
